### PR TITLE
rgw: multisite bucket index trace points

### PR DESCRIPTION
**Pull request by yuvalif**
## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/rgw_obj_types.h [hunk 1]
```diff
@@ -487,6 +488,7 @@ inline std::ostream& operator<<(std::ostream& out, const rgw_raw_obj& o) {
 struct rgw_obj {
   rgw_bucket bucket;
   rgw_obj_key key;
+  jspan_context trace_ctx{false, false};
```
Comment (cbodley) [inline src/rgw/rgw_obj_types.h:491]: `rgw_obj` just names an object. please don't add a trace to it

File: src/rgw/rgw_sal.h [hunk 2]
```diff
@@ -1109,6 +1109,11 @@ class Object {
     /** Get a unique copy of this object */
     virtual std::unique_ptr<Object> clone() = 0;
 
+    /** get the trace stored in the object */
+    virtual jspan_context& get_trace() = 0;
+    /** set a trace to be stored in the object */
+    virtual void set_trace (jspan_context&& _trace_ctx) = 0;
+
```
Comment (cbodley) [inline src/rgw/rgw_sal.h:1116]: i would really prefer not to put this in `sal::Object`. please just pass the span into whichever operations require it. for `UpdateIndex`, that's probably just `sal::Writer::complete()`

File: src/rgw/driver/rados/rgw_sync_module.h [hunk 3]
```diff
@@ -33,7 +33,8 @@ class RGWDataSyncModule {
                                     rgw_bucket_sync_pipe& sync_pipe, rgw_obj_key& key,
                                     std::optional<uint64_t> versioned_epoch,
                                     const rgw_zone_set_entry& my_trace_entry,
-                                    rgw_zone_set *zones_trace) = 0;
+                                    rgw_zone_set *zones_trace,
+                                    const jspan_context *trace_ctx = nullptr) = 0;
```
Comment (cbodley) [inline src/rgw/driver/rados/rgw_sync_module.h:37]: please avoid default arguments in virtual functions

File: src/rgw/driver/rados/rgw_data_sync.cc [hunk 4]
```diff
@@ -4916,6 +4928,8 @@ class RGWBucketShardIncrementalSyncCR : public RGWCoroutine {
 int RGWBucketShardIncrementalSyncCR::operate(const DoutPrefixProvider *dpp)
 {
   int ret;
+  jspan trace;
+  jspan_context trace_ctx{false, false};
```
Comment (cbodley) [inline src/rgw/driver/rados/rgw_data_sync.cc:4932]: it looks like `trace_ctx` is unused

for `trace`, this function is a stackless coroutine, meaning that the stack goes away each time it suspends. i think we should only construct `trace` in the scope just before `spawn(new SyncCR(...`, and move it into that class. that would extend the span's lifetime until the object's sync completes

File: src/rgw/driver/rados/rgw_rados.cc [hunk 5]
```diff
@@ -3912,6 +3915,11 @@ int RGWRados::fetch_remote_obj(RGWObjectCtx& obj_ctx,
   set_mtime_weight.high_precision = high_precision_time;
   int ret;
 
+  auto trace = tracing::rgw::tracer.add_span("fetch_remote_obj", dest_obj.get_trace());
+  trace->SetAttribute(tracing::rgw::OBJECT_NAME, dest_obj.key.name);
+  trace->SetAttribute(tracing::rgw::BUCKET_NAME, dest_bucket_info.bucket.name);
```
Comment (cbodley) [inline src/rgw/driver/rados/rgw_rados.cc:3920]: the trace should probably include the object's instance id and the bucket's tenant

File: src/cls/rgw/cls_rgw_types.cc [hunk 6]
```diff
@@ -876,3 +878,21 @@ std::ostream& operator<<(std::ostream& out, cls_rgw_reshard_status status) {
 
   return out;
 }
+
+void encode_json(const char *name, const jspan_context& span_ctx, Formatter *f) {
+  using namespace opentelemetry;
+  using namespace trace;
```
Comment (cbodley) [inline src/cls/rgw/cls_rgw_types.cc:884]: these namespaces don't look necessary?

**Review by cbodley — COMMENTED** [review]
the cls_rgw bits look great

**Comment by yuvalif** [comment]
unstale

**Comment by yuvalif** [comment]
unstale

**Comment by yuvalif** [comment]
unstale please

**Comment by yuvalif** [comment]
unstale please

**Comment by yuvalif** [comment]
@adamemerson should we close this PR in favor of: https://github.com/ceph/ceph/pull/59365/commits/df096540135327476f470d723ceeb34324b9d6c1 ?

**Comment by yuvalif** [comment]
> @adamemerson should we close this PR in favor of: [df09654](https://github.com/ceph/ceph/commit/df096540135327476f470d723ceeb34324b9d6c1) ?

@adamemerson ?

**Comment by yuvalif** [comment]
closing in favor of: [df09654](https://github.com/ceph/ceph/commit/df096540135327476f470d723ceeb34324b9d6c1)

**Pull request closed without merging**